src/d3a/models/myco_matcher/external_matcher.py

```diff
@@ -74,10 +72,10 @@ def match_recommendations(self, message):
         recommendations = data.get("recommended_matches", [])
         try:
             validated_records = self._get_validated_bid_offer_match_list(recommendations)
-            for market_id, records in validated_records.items():
-                market = self.markets_mapping.get(market_id)
+            for record in validated_records:
+                market = self.markets_mapping.get(record["market_id"])
```

Comment:
Hmm, neither here nor in the other PR I do see the `self.markets_mapping` being populated. Can you help me please?
Also (this should have been commented on by me earlier, sorry):
Why not populating a `market: BifOfferPair` mapping in the frame of `_get_validated_bid_offer_match_list` as we are looping over the recommendations and getting the market objects there already?
This way we could do something like the following here:
``` python
for market, validated_records in self.market_record_mapping.items():
        try:
            market.match_recommendation(records)
        except (OfferNotFoundException, BidNotFoundException):
            # If the offer or bid have just been consumed
            continue
```

Reply:
Thanks a lot, this is indeed a better approach!
I will adapt the _get_validated_bid_offer_match_list function then

Reply:
Actually I gave it another look and recalled that the markets_mapping is cached in publish_offers_bids function, this is before sending the open offers and bids to the myco

Reply:
ok, even better, why not populating the `market: BifOfferPair` there (converting the `markets_mapping` into such).
This way we can match per market here (as suggested).

Reply:
We cannot build bid offer pairs before sending the offers and bids, this is myco-api-client role